src/PyProxyHandler.cc

```diff
@@ -176,29 +176,806 @@ bool PyBaseProxyHandler::isExtensible(JSContext *cx, JS::HandleObject proxy,
+
+  if (!JS_CallFunctionName(cx, selfCopy, fName, jArgs, &rval)) {
+    return false;
+  }
+
+  args.rval().set(rval);
+
+  return true;
+}
+
+static bool array_includes(JSContext *cx, unsigned argc, JS::Value *vp) {
+  return array_copy_func(cx, argc, vp, "includes");
+}
+
+static bool array_forEach(JSContext *cx, unsigned argc, JS::Value *vp) {
+  return array_copy_func(cx, argc, vp, "forEach");
+}
+
+static bool array_map(JSContext *cx, unsigned argc, JS::Value *vp) {
+  return array_copy_func(cx, argc, vp, "map");
+}
+
+static bool array_filter(JSContext *cx, unsigned argc, JS::Value *vp) {
+  return array_copy_func(cx, argc, vp, "filter");
+}
+
+static bool array_reduce(JSContext *cx, unsigned argc, JS::Value *vp) {
+  return array_copy_func(cx, argc, vp, "reduce");
+}
+
+static bool array_reduceRight(JSContext *cx, unsigned argc, JS::Value *vp) {
+  return array_copy_func(cx, argc, vp, "reduceRight");
+}
+
+static bool array_some(JSContext *cx, unsigned argc, JS::Value *vp) {
+  return array_copy_func(cx, argc, vp, "some");
+}
+
+static bool array_every(JSContext *cx, unsigned argc, JS::Value *vp) {
+  return array_copy_func(cx, argc, vp, "every");
+}
+
+static bool array_find(JSContext *cx, unsigned argc, JS::Value *vp) {
+  return array_copy_func(cx, argc, vp, "find");
+}
+
+static bool array_findIndex(JSContext *cx, unsigned argc, JS::Value *vp) {
+  return array_copy_func(cx, argc, vp, "findIndex");
+}
+
+static bool array_flat(JSContext *cx, unsigned argc, JS::Value *vp) {
+  return array_copy_func(cx, argc, vp, "flat", false);
+}
+
+static bool array_flatMap(JSContext *cx, unsigned argc, JS::Value *vp) {
+  return array_copy_func(cx, argc, vp, "flatMap");
+}
+
+static bool array_join(JSContext *cx, unsigned argc, JS::Value *vp) {
+  return array_copy_func(cx, argc, vp, "join", false);
+}
+
+static bool array_toString(JSContext *cx, unsigned argc, JS::Value *vp) {
+  return array_copy_func(cx, argc, vp, "toString", false);
+}
+
+static bool array_toLocaleString(JSContext *cx, unsigned argc, JS::Value *vp) {
+  return array_copy_func(cx, argc, vp, "toLocaleString", false);
+}
+
+static bool array_valueOf(JSContext *cx, unsigned argc, JS::Value *vp) {
+  JS::CallArgs args = JS::CallArgsFromVp(argc, vp);
+
+  JS::RootedObject proxy(cx, JS::ToObject(cx, args.thisv()));
+  if (!proxy) {
+    return false;
+  }
+  PyObject *self = JS::GetMaybePtrFromReservedSlot<PyObject>(proxy, PyObjectSlot);
+
+  // return ref to self
+  args.rval().set(jsTypeFactory(cx, self));
+  return true;
+}
+
+static bool array_entries(JSContext *cx, unsigned argc, JS::Value *vp) {
+  return array_copy_func(cx, argc, vp, "entries", false);
+}
+
+/*  see note in tests
+   static bool array_keys(JSContext *cx, unsigned argc, JS::Value *vp) {
+   return array_copy_func(cx, argc, vp, "keys", false);
+   }
+ */
+
+static bool array_values(JSContext *cx, unsigned argc, JS::Value *vp) {
+  return array_copy_func(cx, argc, vp, "values", false);
+}
```

Comment:
See above comment.

Reply:
It seems it does not apply like above when there is no mutation possible.

Indeed in this situation working on a copy is valid.

Shall we do that for the entries, keys and values methods which cannot have side-effects? It saves a bunch of coding, so is the performance impact minimal? It would typically be called once while getting ready to loop through the result of these method calls

Reply:
Wes may be a better person to answer this question as he would have a better understanding of the Spidermonkey performance implications, but imo creating a copy of a python list as a JS Array every time one of these methods is called could potentially be quite expensive, particularly in the context of DCP where a python list could be being used as a (potentially extremely large) input set to a DCP job. One of the reasons why PythonMonkey was started in the first place is precisely to prevent this kind of translate-by-copying behaviour that Bifrost uses.

Reply:
done